transport/tcp/serve_tcp_ipmap.go

```diff
@@ -45,22 +46,28 @@ func (m *ipMap) Get(ipAddr *net.IPAddr) (string, error) {
 		if e.ip.Equal(ipAddr.IP) {
 			return e.ident, nil
 		}
-		if e.subnet != nil {
-			if e.subnet.Contains(ipAddr.IP) {
-				return strings.Replace(e.ident, "*", ipAddr.IP.String(), 1), nil
+		if e.subnet != nil && e.subnet.Contains(ipAddr.IP) {
+			// If a Zone exists we append it after the IP using a "-" because "%"
+			// is a zfs dataset forbidden char.
+			if e.zone != "" {
+				return strings.Replace(e.ident, "*", ipAddr.IP.String()+"-"+e.zone, 1), nil
 			}
+			// Because of actions taken earlier Line #84 Subnets will
+			// only contain 1 "*" for replacement.
+			return strings.Replace(e.ident, "*", ipAddr.IP.String(), 1), nil
```

Comment:
I think we should incorporate the logic in the if branch into a separate helper function, e.g. `zfsDatasetPathComponentCompatibleRepresentation(addr *net.IpAddr)`